drivers/sensor/grow_r502a/grow_r502a.c

```diff
@@ -87,41 +157,139 @@ static void uart_cb_handler(const struct device *dev, void *user_data)
 
 		while (uart_irq_rx_ready(dev)) {
 			len = uart_fifo_read(dev, &drv_data->rx_buf.data[offset],
-						R502A_BUF_SIZE - offset);
+								drv_data->pkt_len);
 			offset += len;
 			drv_data->rx_buf.len = offset;
 
-			if (offset >= R502A_HEADER_LEN) {
-				pkt_sz = R502A_HEADER_LEN +
-						drv_data->rx_buf.data[R502A_HEADER_LEN-1];
+			if (drv_data->pkt_len != len) {
+				drv_data->pkt_len -= len;
+				continue;
 			}
-			if (offset < pkt_sz) {
+
+			if (offset == R502A_HEADER_LEN) {
+				drv_data->pkt_len = sys_get_be16(
+							&drv_data->rx_buf.data[R502A_PKG_LEN_IDX]
+							);
 				continue;
 			}
+
 			LOG_HEXDUMP_DBG(drv_data->rx_buf.data, offset, "RX");
+			uart_irq_rx_disable(dev);
 			k_sem_give(&drv_data->uart_rx_sem);
 			break;
 		}
 	}
 }
 
-static int fps_led_control(const struct device *dev, struct led_params *led_control)
+/**
+ * @brief	Set sensor device's basic parameters like baud rate, security level
+ *		and data package length.
+ */
+static int fps_set_sys_param(const struct device *dev, const struct sensor_value *val)
+{
+	struct grow_r502a_data *drv_data = dev->data;
+	union r502a_packet rx_packet = {0};
+	int ret = 0;
+	char const set_sys_param_len = 3;
+
+	union r502a_packet tx_packet = {
+		.pid = R502A_COMMAND_PACKET,
+		.data = { R502A_SETSYSPARAM, val->val1, val->val2}
+	};
+
+	k_mutex_lock(&drv_data->lock, K_FOREVER);
+
+	ret = transceive_packet(dev, &tx_packet, &rx_packet, set_sys_param_len);
+	if (ret != 0) {
+		goto unlock;
+	}
+
+	ret = r502a_validate_rx_packet(&rx_packet);
+	if (ret != 0) {
+		goto unlock;
+	}
+
+	if (rx_packet.buf[R502A_CC_IDX] == R502A_OK) {
+		LOG_DBG("R502A set system parameter success");
+	} else {
+		LOG_ERR("R502A set system parameter error %d", rx_packet.buf[R502A_CC_IDX]);
+		ret = -EIO;
+		goto unlock;
+	}
+
+unlock:
+	k_mutex_unlock(&drv_data->lock);
+	return ret;
+}
+
+static int fps_read_sys_param(const struct device *dev, struct r502a_sys_param *val)
+{
+	struct grow_r502a_data *drv_data = dev->data;
+
+	union r502a_packet rx_packet = {0};
+	int offset = 0, ret = 0;
+	char const read_sys_param_len = 1;
+
+	union r502a_packet tx_packet = {
+		.pid = R502A_COMMAND_PACKET,
+		.data = {R502A_READSYSPARAM}
+	};
+
+	k_mutex_lock(&drv_data->lock, K_FOREVER);
+
+	ret = transceive_packet(dev, &tx_packet, &rx_packet, read_sys_param_len);
+	if (ret != 0) {
+		goto unlock;
+	}
+
+	ret = r502a_validate_rx_packet(&rx_packet);
+	if (ret != 0) {
+		goto unlock;
+	}
+
+	if (rx_packet.buf[R502A_CC_IDX] == R502A_OK) {
+		LOG_DBG("R502A read system parameter success");
+	} else {
+		LOG_ERR("R502A read system parameter error %d", rx_packet.buf[R502A_CC_IDX]);
+		ret = -EIO;
+		goto unlock;
+	}
+
+	val->status_reg = sys_get_be16(&rx_packet.data[1]);
```

Comment:
use offsetof for all the index

Reply:
addressed in the latest changes. please have a look